File: 基于最近邻超圆盘分类的CWRU轴承数据集故障识别/make_dataset.m
function [segmented_data, Class_] = make_dataset(data_src, num_samples, class_)

    % 获取目标文件夹数据
    find_file = fullfile(data_src, '/*.mat');
    files = dir(find_file);
    files = sort({files.name});
    data = load(fullfile(data_src, files{1})); % struct数据类型
    keysList = fieldnames(data);
    my_key = '';
    for i = 1:length(keysList)
        if contains(keysList{i}, '_DE_time')
            my_key = keysList{i};
            break;
        end
    end
    drive_end_data = data.(my_key);
    drive_end_data = drive_end_data(:);

    % 分割数据块
    num_segments = floor(length(drive_end_data)/num_samples);
    slices = drive_end_data(1:num_segments*num_samples);
    segmented_data = reshape(slices, [num_samples, num_segments])'; % matlab的reshape是按列依此重组的

    % 预防性操作
    files = files(2:end);
    for i = 1:length(files)
        data = load(fullfile(data_src, files{i}));
        keysList = fieldnames(data);
        for j = 1:length(keysList)
            if contains(keysList{j}, '_DE_time')
                my_key = keysList{j};
            end
        end
        drive_end_data = data.(my_key);
        drive_end_data = drive_end_data(:);
        num_segments = floor(length(drive_end_data)/num_samples);
        slices = drive_end_data(1:num_segments*num_samples);
        segmented_data = [segmented_data; reshape(slices, [num_samples, num_segments])'];
    end
    
    segmented_data = unique(segmented_data, 'rows'); % remove duplicates
    segmented_data = segmented_data(randperm(size(segmented_data, 1)), :); % shuffle the data
    Class_ = ones(size(segmented_data, 1), 1) * class_;

end

File: 基于最近邻超圆盘分类的CWRU轴承数据集故障识别/make_dataset.py
import numpy as np
from scipy.io import loadmat
import glob
import re
import os

def make_dataset(data_src, num_samples, class_):
    
    # 获取目标文件夹数据
    req_key = "_DE_time" 
    pattern = re.compile(req_key)
    files = glob.glob(data_src)
    files = np.sort(files)
    data = loadmat(files[0])
    keysList = [key for key in data]
    for key in keysList:
        if pattern.search(key):
            my_key = key
    drive_end_data = data[my_key]
    drive_end_data = drive_end_data.reshape(-1) # 变一维
    # 分割数据块
    num_segments = np.floor(len(drive_end_data)/num_samples)
    #slices = np.split(drive_end_data[0:int(num_segments*num_samples)], num_samples)
    slices = drive_end_data[0:int(num_segments*num_samples)]
    silces = slices.reshape(int(num_segments), num_samples)
    segmented_data = silces
    # 预防性操作
    files = files[1:]
    for file in files:
        data = loadmat(file)
        keysList = [key for key in data]
        for key in keysList:
            if pattern.search(key):
                my_key = key
        drive_end_data = data[my_key]
        drive_end_data = drive_end_data.reshape(-1)
        num_segments = np.floor(len(drive_end_data)/num_samples)
        #slices = np.split(drive_end_data[0:int(num_segments*num_samples)], num_samples)
        slices = drive_end_data[0:int(num_segments*num_samples)]
        silces = slices.reshape(int(num_segments), num_samples)
        segmented_data = np.concatenate( (segmented_data, silces) , axis=0, out=None)
    
    segmented_data = np.unique(segmented_data, axis= 0) # remove duplicates
    np.random.shuffle( segmented_data) # suffule the data
    Class_ = np.ones(len(segmented_data)).astype(int)*class_
    
    return segmented_data, Class_